src/ipc_listener.cpp

```diff
@@ -208,9 +260,13 @@ int zmq::ipc_listener_t::set_address (const char *addr_)
 
 int zmq::ipc_listener_t::close ()
 {
-    zmq_assert (s != retired_fd);
-    int rc = ::close (s);
-    errno_assert (rc == 0);
+    int rc;
+
+    if ( s != -1 ) {
+      zmq_assert (s != retired_fd);
```

Comment:
> I may need to rework it a bit.  What I was trying to do was cleanup
> the mkdtemp() on an error.  I thought that re-using
> ipc_listener_t::close() was the best way to do that.

If you need more time, I can merge as-is for now, and then you can send another PR to fix it when it's ready. Or maybe it's easier for you to work on the same branch?

> I can revert ipc_listener_t::close(), and update
> ipc_listener_t::set_address() to properly cleanup the directory if
> open_socket() fails.  I understand your API checks, and I'll fix
> things on my end.
> 
> Are the prohibitions on the use of exceptions?  I initially coded it
> to throw an exception in create_wildcard_address(), but didn't see
> handling of exceptions elsewhere.  If exceptions are ok, I can rework
> set_address() to make sure that there are no stray files or
> directories upon failure pretty easily.  Otherwise, the code may look
> a bit ugly.

I personally don't have any objections to exceptions, but it's true that they are not used anywhere else. @hintjens any objections/opinions on exceptions?